presto-main/src/main/java/com/facebook/presto/tz/JdkBasedZoneInfoProvider.java

```diff
@@ -0,0 +1,84 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.tz;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.joda.time.DateTimeZone;
+import org.joda.time.tz.CachedDateTimeZone;
+import org.joda.time.tz.Provider;
+
+import java.time.zone.ZoneRulesProvider;
+import java.util.Set;
+
+public class JdkBasedZoneInfoProvider
+        implements Provider
+{
+    private final LoadingCache<String, DateTimeZone> zoneCache;
+
+    {
+        // It is necessary to cache the string to zone mapping.
+        // Otherwise, the zone-internal cache won't be useful.
+        // Joda's default Provider implementation, ZoneInfoProvider, caches with SoftReference.
+        // This implementation didn't use SoftReference for simplicity.
+        zoneCache = CacheBuilder.newBuilder().build(new CacheLoader<String, DateTimeZone>()
+            {
+                @Override
+                public DateTimeZone load(String id)
+                {
+                    DateTimeZone zone = new JdkBasedDateTimeZone(id);
+                    if (!zone.isFixed()) {
+                        // Joda's default Provider implementation, ZoneInfoProvider, caches a zone
+                        // most of the time if the zone is not fixed. The exception being zones that
+                        // isn't worthwhile from a performance perspective. The actual check is in
+                        // `DateTimeZoneBuilder.isCachable` check (a zone would fail if on average
+                        // it has a transition less than every 25 days).
+                        // This implementation caches everything for simplicity.
+                        // Note that this is cache within a zone, which is different from the zone id
+                        // to zone cache.
+                        zone = CachedDateTimeZone.forZone(zone);
+                    }
+                    return zone;
+                }
+            });
+        // Joda requires that this particular zone implementation be used for UTC.
+        zoneCache.put("UTC", DateTimeZone.UTC);
+    }
+
+    @Override
+    public DateTimeZone getZone(String id)
+    {
+        if (id.startsWith("+") || id.startsWith("-")) {
+            // DateTimeZone.forID takes care of this case after this method returns null
+            return null;
+        }
+        return zoneCache.getUnchecked(id);
+    }
+
+    @Override
+    public Set<String> getAvailableIDs()
+    {
+        return ZoneRulesProvider.getAvailableZoneIds();
+    }
+
+    public static void registerAsJodaZoneInfoProvider()
```

Comment:
If you set it through `setProvider`, you won't be able to tell whether any one has used DateTimeZone before it's set. I think the system property way is better because it makes it possible to validate that.

Reply:
Add a comment explaining that so we don't end up discussing this again in the future.